docs/proposals/improve-addons.md

```diff
@@ -0,0 +1,350 @@
+- Generic Pull from object storage. It is definitely a feature that should be on roadmap for addons after the implementation of this proposal is done and should fit right in the current architecture. We have made this decision after consulting with PS. It was decided to go implement Git as an external source first, as this is going to reach the majority of customers
+
+# Implementation
+
+## Architecture
+
+See below for a high level architecture view. Each component will be explained in detail in the subsequent chapters of this proposal.
+
+![Architecture](./images/improveaddons-architecture.png)
+
+([source](https://app.diagrams.net/#G1drryGHt2MbCDY6wRz99u7lMSBYXgI2lH))
+
+## Components
+
+### AddonCatalogueDistributor
+
+The AddonCatalogueDistributor resides in the master cluster. Its main job is to merge different AddonCatalogues and push an aggregated AddonCatalogue into different seed clusters. We have decided to create a separate component for this, so there is a single source of truth on available addons, which can be synced across multiple seed clusters.
+
+In order to work properly, the AddonCatalogueDistributor must (dynamically) be able to discover all available SeedClusters, so it knows where to push the catalogue to.
+
+In the case of the MasterCluster and SeedCluster being the same, we propose to still run the AddonCatalogueDistributor exactly the same. Concretely this means it would have one target seed cluster (itself) and create a merged catalogue for it. This would have the advantage, that in case customers decide to add new seed clusters, the new seed clusters would work right out-of-the-box. Additionally, this helps keeping the code generic, which is in line with KKPs development pattern.
+
+In the final implementation, we propose to have at least two catalogues per KKP installation. We think this makes sense to protect core addons, which are essential for all clusters, to be protected from third-party influence:
+
+- CoreAddonCatalogue → will be maintained by Kubermatic. Should not be directly edited by the customer. We think it makes sense to add this as a layer of protection against misconfiguration. This catalogue is rarely changed. Moreover, this catalogue will not be shown in the UI. This catalogue contains addons for the supported Kubernetes versions (eg KubeProxy will have 3 versions of the addons, one per K8s version. OpenVPN will have one version of the addons compatible with the 3 K8s versions)
+- CustomAddonCatalogue(s) →one or multiple catalogues that contain custom addons. This catalogue can be changed more often. Additionally, it is possible to have multiple CustomAddonCatalogues
+
+Furthermore, the AddonCatalogueDistributor ensures that there are no duplicate addons inside the AddonCatalogue. A duplicate is defined by two addons having the same `name` and `version`. In these situations, we propose the AddonCatalogueDistributor to stop its distribution and log an error.
```

Comment:
I'd suggest to validate the `AddonCatalogue` resources through a webhook instead, so it gets rejected and doesn't take down the AddonCatalogueDistributor as a whole.

Reply:
Ah interesting. Maybe I misunderstood our current implementation. I thought our current implementation would do exactly that: ensure that CRDs and namespaces are deployed before any other manifests.
cc @vgramer: This is something I think we have to look in the code of the current implementation. But yes, if the current system is not able to install CRDs before deploying resources, then we should include that capability in our proposal.

On the other side: For the helm rendering method, handling dependencies is done directly via helm; So theoretically this should come out of the box (of course during implementation we will have to see how well it works in practice)

Reply:
Not sure we're talking about the same thing here :D I was referring to the duplicate detection in an (instance of) `AddonCatalogue`. Maybe I misunderstood this part of the proposal, but it looked like the validation of resources was supposed to happen in the distributor, while a webhook would help by rejecting invalid `AddonCatalogues` before they hit the cluster and can be processed by the AddonCatalogueDistributor.

Reply:
I commented on the wrong comment on accident. I have moved my comment since then. Will get back to you about the webhook soon. Just have to think about it for a minute